tests/notifications.rs

```diff
@@ -0,0 +1,125 @@
+use std::{
+  sync::Arc,
+  time::Duration,
+};
+
+use async_trait::async_trait;
+
+use tokio::{
+  self,
+  test as atest,
+  sync::{
+    mpsc::{UnboundedSender, unbounded_channel},
+    Notify,
+  },
+  process::{ChildStdin, Command},
+  task::yield_now,
+};
+
+use futures::FutureExt;
+
+use nvim_rs::{
+  self,
+  create::tokio as create,
+  compat::tokio::Compat,
+  neovim::Neovim,
+  Value,
+};
+
+mod common;
+use common::*;
+
+const ITERS: usize = 25;
+const TIMEOUT: Duration = Duration::from_secs(60);
+
+macro_rules! timeout {
+  ($x:expr) => {
+    tokio::time::timeout(TIMEOUT, $x).map(|res| {
+      res.expect(&format!("Timed out waiting for {}", stringify!($x)))
+    })
+  }
+}
+
+#[derive(Clone)]
+struct Handler {
+  result_sender: UnboundedSender<u8>,
+  notifiers: Arc<Vec<Notify>>,
+}
+
+#[async_trait]
+impl nvim_rs::Handler for Handler {
+  type Writer = Compat<ChildStdin>;
+
+  async fn handle_notify(
+    &self,
+    name: String,
+    args: Vec<Value>,
+    _: Neovim<Self::Writer>,
+  ) {
+    assert_eq!(name, "idx");
+    let idx = args[0].as_i64().unwrap();
+
+    /* Wait until we've received a message from the test, then send back the
+     * number we received.
+     */
+    self.notifiers[idx as usize].notified().await;
+    self.result_sender.send(idx as u8).unwrap();
+  }
+}
+
+#[atest]
+async fn sequential_notifications() {
+  // Create a tokio notifier for each nvim notification that we'll be handling
+  let mut notifiers = Vec::<Notify>::with_capacity(ITERS);
+  for _ in 0..ITERS {
+    notifiers.push(Notify::new());
+  }
+  let notifiers = Arc::new(notifiers);
+
+  /* Create a channel the notification handler will use each time for writing
+   * back the number provided by each notification
+   */
+  let (result_sender, mut result_receiver) = unbounded_channel();
+  let handler = Handler {
+    result_sender,
+    notifiers: notifiers.clone(),
+  };
+
+  // Generate the commands to send the notifications
+  let mut nvim_cmds = Vec::<String>::with_capacity(ITERS);
+  for i in 0..ITERS {
+    nvim_cmds.push(format!("call rpcnotify(1, 'idx', {})", i));
+  }
+
+  // Startup nvim
+  let (_nvim, _io_handler, _child) = create::new_child_cmd(
+    Command::new(nvim_path()).args(&[
+      "-u",
+      "NONE",
+      "--embed",
+      "--headless",
+      "-c",
+      nvim_cmds.join("|").as_str(),
+    ]),
+    handler
+  )
+  .await
+  .unwrap();
+
+  /* Unblock notifications in the reverse order that they were sent, if
+   * notifications are being handled sequentially then they should still be
+   * processed in their original order.
+   */
+  for notifier in notifiers.iter().rev() {
+    notifier.notify_one();
+    /* Yield once, so that we guarantee the notification handler for this
+     * notification executes pre-maturely if it was handled out of order.
+     */
+    yield_now().await;
+  }
+
+  // Check the results
+  for i in 0..ITERS {
+    assert_eq!(i as u8, timeout!(result_receiver.recv()).await.unwrap());
+  }
+}
```

Comment:
Sheesh-the one thing I forgot to do before submitting this :S, this test doesn't seem to actually work like I'd hope so I'll have to take another look at this

Reply:
The problem might be that it's not easy to make notification handling out-of-order even in the old implementation. It seemed only to be noticable on systems under load, which is kinda hard to emulate. Maybe add some task sleeping?

Reply:
Yeah-that seems to help. Something I realized as well though, which might let us still avoid using sleeps is- we don't technically know when nvim has finished running the commands we've passed to it. So it's entirely possible that by the time we've set off all of the notifiers, neovim hasn't even run anything in `-c` yet. probably could actually fix this by setting up a second rpc channel, and sending a single notification to it at the end so we know that neovim has finished sending us the notifications. I'll try that, but if that doesn't work I can just use some sleeps

Reply:
I ended up being able to fix this pretty nicely by just using `nvim.command(…)` for sending the rpc notifications instead of trying to pass them via `-c`. Since the callback doesn't return until neovim has finished executing the commands, that ensures that the notifications are in the queue.
Will continue working through the rest of the comments here :)